rollup/internal/controller/permissionless_batches/minimal_recovery.go

```diff
@@ -0,0 +1,367 @@
+package permissionless_batches
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/scroll-tech/da-codec/encoding"
+	"github.com/scroll-tech/go-ethereum/core"
+	"github.com/scroll-tech/go-ethereum/ethclient"
+	"github.com/scroll-tech/go-ethereum/log"
+	"github.com/scroll-tech/go-ethereum/rollup/l1"
+	"gorm.io/gorm"
+
+	"scroll-tech/common/types"
+	"scroll-tech/database/migrate"
+	"scroll-tech/rollup/internal/config"
+	"scroll-tech/rollup/internal/controller/watcher"
+	"scroll-tech/rollup/internal/orm"
+)
+
+const (
+	// defaultRestoredChunkIndex is the default index of the last restored fake chunk. It is used to be able to generate new chunks pretending that we have already processed some chunks.
+	defaultRestoredChunkIndex uint64 = 1337
+	// defaultRestoredBundleIndex is the default index of the last restored fake bundle. It is used to be able to generate new bundles pretending that we have already processed some bundles.
+	defaultRestoredBundleIndex uint64 = 1
+)
+
+type MinimalRecovery struct {
+	ctx       context.Context
+	cfg       *config.Config
+	genesis   *core.Genesis
+	db        *gorm.DB
+	chunkORM  *orm.Chunk
+	batchORM  *orm.Batch
+	bundleORM *orm.Bundle
+
+	chunkProposer  *watcher.ChunkProposer
+	batchProposer  *watcher.BatchProposer
+	bundleProposer *watcher.BundleProposer
+	l2Watcher      *watcher.L2WatcherClient
+}
+
+func NewRecovery(ctx context.Context, cfg *config.Config, genesis *core.Genesis, db *gorm.DB, chunkProposer *watcher.ChunkProposer, batchProposer *watcher.BatchProposer, bundleProposer *watcher.BundleProposer, l2Watcher *watcher.L2WatcherClient) *MinimalRecovery {
+	return &MinimalRecovery{
+		ctx:            ctx,
+		cfg:            cfg,
+		genesis:        genesis,
+		db:             db,
+		chunkORM:       orm.NewChunk(db),
+		batchORM:       orm.NewBatch(db),
+		bundleORM:      orm.NewBundle(db),
+		chunkProposer:  chunkProposer,
+		batchProposer:  batchProposer,
+		bundleProposer: bundleProposer,
+		l2Watcher:      l2Watcher,
+	}
+}
+
+func (r *MinimalRecovery) RecoveryNeeded() bool {
+	chunk, err := r.chunkORM.GetLatestChunk(r.ctx)
+	if err != nil {
+		return true
+	}
+	if chunk.Index <= defaultRestoredChunkIndex {
+		return true
+	}
+
+	batch, err := r.batchORM.GetLatestBatch(r.ctx)
```

Comment:
If I understand correctly, the users use it first time, they have a empty db. so they will get null latest chunk/batch/bundle, so these sanity checks maybe not be useful.